hazelcast/src/test/java/com/hazelcast/test/compatibility/SamplingSerializationService.java

```diff
@@ -0,0 +1,255 @@
+        return (T) delegate.readObject(in);
+    }
+
+    @Override
+    public <T> T readObject(ObjectDataInput in, Class aClass) {
+        return (T) delegate.readObject(in, aClass);
+    }
+
+    @Override
+    public <T> T toObject(Object data) {
+        return (T) delegate.toObject(data);
+    }
+
+    @Override
+    public <T> T toObject(Object data, Class klazz) {
+        return (T) delegate.toObject(data, klazz);
+    }
+
+    @Override
+    public byte[] toBytes(Object obj) {
+        byte[] bytes = delegate.toBytes(obj);
+        sampleObject(obj, bytes);
+        return bytes;
+    }
+
+    @Override
+    public byte[] toBytes(int padding, Object obj) {
+        byte[] bytes = delegate.toBytes(padding, obj);
+        sampleObject(obj, bytes);
+        return bytes;
+    }
+
+    @Override
+    public byte[] toBytes(Object obj, PartitioningStrategy strategy) {
+        byte[] bytes = delegate.toBytes(obj, strategy);
+        sampleObject(obj, bytes);
+        return bytes;
+    }
+
+    @Override
+    public byte[] toBytes(int padding, Object obj, PartitioningStrategy strategy) {
+        byte[] bytes = delegate.toBytes(padding, obj, strategy);
+        sampleObject(obj, bytes);
+        return bytes;
+    }
+
+    @Override
+    public ClassLoader getClassLoader() {
+        return delegate.getClassLoader();
+    }
+
+    @Override
+    public BufferObjectDataInput createObjectDataInput(byte[] data) {
+        return delegate.createObjectDataInput(data);
+    }
+
+    @Override
+    public BufferObjectDataInput createObjectDataInput(Data data) {
+        return delegate.createObjectDataInput(data);
+    }
+
+    @Override
+    public BufferObjectDataOutput createObjectDataOutput(int size) {
+        return delegate.createObjectDataOutput(size);
+    }
+
+    @Override
+    public ManagedContext getManagedContext() {
+        return delegate.getManagedContext();
+    }
+
+    @Override
+    public PortableReader createPortableReader(Data data)
+            throws IOException {
+        return delegate.createPortableReader(data);
+    }
+
+    @Override
+    public PortableContext getPortableContext() {
+        return delegate.getPortableContext();
+    }
+
+    @Override
+    public void disposeData(Data data) {
+        delegate.disposeData(data);
+    }
+
+    @Override
+    public BufferObjectDataOutput createObjectDataOutput() {
+        return delegate.createObjectDataOutput();
+    }
+
+    @Override
+    public ByteOrder getByteOrder() {
+        return delegate.getByteOrder();
+    }
+
+    @Override
+    public byte getVersion() {
+        return delegate.getVersion();
+    }
+
+    @Override
+    public void dispose() {
+        delegate.dispose();
+    }
+
+    // record the given object, then return it
+    private static <T> T sampleObject(T obj, byte[] serializedObject) {
+        if (obj == null)
+            return null;
+
+        if (serializedObject != null && shouldAddSerializedSample(obj)) {
```

Comment:
> If I call sampleObject method twice. Is not second call will skip
addSerializedSample(obj, serializedObject); (line 203),
because SAMPLED_CLASSES.contains(className) check will hold true in line 224 ?

No, unless `MAX_SERIALIZED_SAMPLES_PER_CLASS` have already been collected. In the first invocation to `shouldAddSerializedSample` the flow will be as follows:
- L219 `if (klass.isPrimitive())`, continue execution
- L224 `if (SAMPLED_CLASSES.contains(className))` `false`, continue execution
- ...
- L237 `if (existingSamples == null)` is `true` --> returns `true` and sample if picked up
Notice, the class name was not added to  `SAMPLED_CLASSES`.

Second invocation, is same as above but:
- L237 `if (existingSamples == null)` is `false` --> continue
- L240 `if (existingSamples.size() < MAX_SERIALIZED_SAMPLES_PER_CLASS)` is `true` --> return true and sample is collected, class name still not added to `SAMPLED_CLASSES`.

Only when the number of collected classes is at least `MAX_SERIALIZED_SAMPLES_PER_CLASS` does L240 continue further to execute
```
SAMPLED_CLASSES.add(className);
return false;
```